src/Hedgehog/Plutus/Tx.hs

```diff
@@ -106,11 +131,140 @@ data ScriptPurpose
 
   In the future, hps should support custom balancers.
 -}
-balanceTx :: Model.Mock -> Tx 'Unbalanced -> Maybe (Hedgehog.Gen (Tx 'Balanced))
-balanceTx = _
-
-confirmBalanced :: Model.Mock -> Tx 'Unbalanced -> Maybe (Tx 'Balanced)
-confirmBalanced = _
+balanceTx :: TxContext -> Tx 'Unbalanced -> Maybe (Hedgehog.Gen (Tx 'Balanced))
+balanceTx context tx = balanceTxWhere context tx (const True) (const True)
+
+-- | as balanceTx but only uses txOuts belonging to a particular PubKeyHash and always sends change to that PubKeyHash
+balanceTxAsPubKey :: TxContext -> Tx 'Unbalanced -> PubKeyHash -> Maybe (Hedgehog.Gen (Tx 'Balanced))
+balanceTxAsPubKey context tx pkh =
+  balanceTxWhere
+    context
+    tx
+    ( Plutus.txOutAddress
+        >>> Plutus.addressCredential
+        >>> ( \case
+                Plutus.PubKeyCredential pkh' -> pkh' == pkh
+                _ -> False
+            )
+    )
+    (== pkh)
+
+-- | as balanceTx but with additional predicates about which TxOuts and PubKeyHashs can be used
+balanceTxWhere ::
+  TxContext ->
+  Tx 'Unbalanced ->
+  (Plutus.TxOut -> Bool) ->
+  (Plutus.PubKeyHash -> Bool) ->
```

Comment:
Do you have a motivation for a filter on the output, or is it just for completeness?

Reply:
Yeah, I needed a filter (or something similar) to make `balanceTxAsPubKey` return change to the expected user.